include/rocksdb/ipp_aes_ctr_provider.h

```diff
@@ -0,0 +1,123 @@
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#ifdef IPPCP
+
+#if !defined(ROCKSDB_LITE)
+
+#include <wmmintrin.h>
+
+#include <string>
+
+#include "env_encryption.h"
+#include "ippcp.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+enum struct KeySize { AES_128 = 16, AES_192 = 24, AES_256 = 32 };
+
+#if defined(HAVE_SSE42) || defined(HAVE_SSE2)
+union IV_t {
+  IV_t() { memset(u8_, 0, sizeof(u8_)); }
+
+  IV_t(const IV_t& rhs) {
+    u64_[0] = rhs.u64_[0];
+    u64_[1] = rhs.u64_[1];
+  }
+
+  IV_t& operator=(const IV_t& rhs) {
+    u64_[0] = rhs.u64_[0];
+    u64_[1] = rhs.u64_[1];
+    return *this;
+  }
+
+  IV_t& operator=(const char* init_vector) {
+    memcpy(u8_, init_vector, sizeof(u8_));
+    return *this;
+  }
+
+  IV_t& operator+=(uint64_t n) {
+    u64_[0] += n;
+    if (u64_[0] == 0) ++u64_[1];
+    return *this;
+  }
+
+  void to_char_array(char* ctr_block) {
+    size_t n = sizeof(u8_);
+    for (size_t i = 0; i < n; ++i) ctr_block[i] = u8_[n - 1 - i];
+  }
+
+ private:
+  uint8_t u8_[16];
+  uint64_t u64_[2];
+};
+#endif
+
+class IppAESCTRCipherStream : public rocksdb::BlockAccessCipherStream {
+ public:
+  static const size_t kBlockSize = 16;    // in bytes
+  static const size_t kCounterLen = 128;  // in bits
+
+  IppAESCTRCipherStream(IppsAESSpec* aes_ctx, const char* init_vector);
+
+  virtual rocksdb::Status Encrypt(uint64_t fileOffset, char* data,
+                                  size_t dataSize) override;
+  virtual rocksdb::Status Decrypt(uint64_t fileOffset, char* data,
+                                  size_t dataSize) override;
+  virtual size_t BlockSize() override { return kBlockSize; }
+
+ protected:
+  virtual void AllocateScratch(std::string&) override {}
+  virtual rocksdb::Status EncryptBlock(uint64_t, char*, char*) override {
+    return rocksdb::Status::OK();
+  }
+  virtual rocksdb::Status DecryptBlock(uint64_t, char*, char*) override {
+    return rocksdb::Status::OK();
+  }
+
+ private:
+  IppsAESSpec* aes_ctx_;
+#if defined(HAVE_SSE42) || defined(HAVE_SSE2)
+  __m128i init_vector_;
+#else
+  IV_t init_vector_;
+#endif
+};
+
+class IppAESCTRProvider : public rocksdb::EncryptionProvider {
+ public:
+  IppAESCTRProvider(const char* key, KeySize key_size);
+  IppAESCTRProvider(const IppAESCTRProvider&) = delete;
+  IppAESCTRProvider& operator=(const IppAESCTRProvider&) = delete;
+
+  virtual size_t GetPrefixLength() override { return 4096; }
+
+  virtual rocksdb::Status CreateNewPrefix(const std::string& fname,
+                                          char* prefix,
+                                          size_t prefixLength) override;
+
+  virtual rocksdb::Status CreateCipherStream(
+      const std::string& fname, const rocksdb::EnvOptions& options,
+      rocksdb::Slice& prefix,
+      std::unique_ptr<rocksdb::BlockAccessCipherStream>* result) override;
+
+  virtual ~IppAESCTRProvider() {
```

Comment:
Can we move the implementation to the C++ file instead?  Then we might not need to include some of the other header files here.